File: services/sales_root_cause.py
from __future__ import annotations

"""Mağaza satış sapmasını mevcut operasyon ve iş gücü kanıtlarıyla açıklar.

Hesaplar UI'dan bağımsız tutulur; Streamlit ve diğer sunucu tüketicileri aynı
mağaza/dönem eşleşmesini kullanabilir. Ana norm veya AI motoruna dokunmaz.
"""

from typing import Any

import pandas as pd


def _number(value: Any) -> float | None:
    try:
        result = float(value)
        return result if pd.notna(result) else None
    except (TypeError, ValueError):
        return None


def _period(frame: pd.DataFrame) -> pd.Series:
    for name in ("Ay", "Dönem", "Donem", "Tarih"):
        if name in frame.columns:
            return frame[name].astype(str).str[:7]
    return pd.Series("", index=frame.index, dtype="object")


def _store_key(frame: pd.DataFrame) -> pd.Series:
    for name in ("MagazaID", "MağazaID", "Mağaza", "Magaza"):
        if name in frame.columns:
            return frame[name].fillna("").astype(str).str.strip()
    return pd.Series("", index=frame.index, dtype="object")


def _indexed(frame: pd.DataFrame | None, period: str = "") -> pd.DataFrame:
    if frame is None or frame.empty:
        return pd.DataFrame()
    view = frame.copy()
    periods = _period(view)
    if period:
        view = view[periods.eq("") | periods.eq(period)].copy()
    view["_store_key"] = _store_key(view)
    return view.drop_duplicates("_store_key", keep="last").set_index("_store_key")


def _value(frame: pd.DataFrame, key: str, *columns: str) -> float | None:
    if frame.empty or key not in frame.index:
        return None
    row = frame.loc[key]
    for column in columns:
        if column in frame.columns:
            return _number(row.get(column))
    return None


def _change(current: float | None, previous: float | None) -> float | None:
    if current is None or previous in (None, 0):
        return None
    return (current / previous - 1) * 100


def _diagnose(row: dict[str, Any]) -> tuple[str, str, str]:
    sales_rate, norm_rate = row.get("Hedef Gerçekleşme %"), row.get("Norm Karşılama %")
    if sales_rate is None:
        return "VERİ EKSİK", "Belirsiz", "Satış hedefi ve gerçekleşen dönem verisini tamamla."
    if sales_rate >= 100 and norm_rate < 95:
        return "EKSİK KADROYLA HEDEF ÜSTÜ", "Desteklenmiyor", "Mesai ve çalışan yükü sürdürülebilirliğini incele."
    if sales_rate >= 100:
        return "HEDEF TUTTU", "Desteklenmiyor", "Fire ve iş yükü göstergelerini koruma sınırı olarak izle."
    if norm_rate >= 98:
        return "TAM KADRO / DÜŞÜK SATIŞ", "Desteklenmiyor", "Fiş, sepet, fire ve mağaza yönetimi açıklaması iste."
    if (row.get("Fiş Değişim %") or 0) < -3 and (row.get("Sepet Değişim %") or 0) >= -3:
        return "MÜŞTERİ TRAFİĞİ DÜŞÜŞÜ", "Desteklenmiyor", "Trafik, kampanya, rekabet ve mağaza çekiciliğini açıkla."
    if (row.get("Sepet Değişim %") or 0) < -3:
        return "ORTALAMA SEPET DÜŞÜŞÜ", "Desteklenmiyor", "Ürün karması, fiyat, kampanya ve stok bulunurluğunu açıkla."
    people_evidence = norm_rate < 95 and (
        (row.get("Fazla Mesai Saat") or 0) > 0
        or (row.get("Kayıp FTE") or 0) > 0
        or (row.get("İş Yükü Endeksi") or 0) >= 70
    )
    if people_evidence:
        return "PERSONEL ETKİSİ KANITLI", "Kanıtlı", "Eksik unvan/saat ile satış kaybı bağlantısını ve telafi planını kaydet."
    return "KÖK NEDEN AÇIKLANMADI", "Belirsiz", "Fiş, sepet, stok, kategori ve yönetici kırılımıyla kanıt sun."


def build_sales_root_cause(
    *, sheets: dict[str, pd.DataFrame], stores: pd.DataFrame,
    targets: list[dict[str, Any]] | pd.DataFrame, inflation_pct: float = 32.11,
) -> tuple[pd.DataFrame, str, str]:
    operations = sheets.get("Aylık Operasyon KPI", sheets.get("Aylik Operasyon KPI", pd.DataFrame())).copy()
    if operations.empty or stores.empty:
        return pd.DataFrame(), "", ""
    periods = sorted(p for p in _period(operations).dropna().unique().tolist() if p)
    latest, previous = (periods[-1] if periods else ""), (periods[-2] if len(periods) > 1 else "")
    current, prior = _indexed(operations, latest), _indexed(operations, previous)
    overtime = _indexed(sheets.get("Fazla Mesai"), latest)
    absence = _indexed(sheets.get("Devamsızlık"), latest)
    workload = _indexed(sheets.get("İş Yükü Endeksi", sheets.get("Is Yuku Endeksi")))
    waste = _indexed(sheets.get("Fire ve İade", sheets.get("Fire ve Iade")), latest)
    performance = _indexed(sheets.get("Performans"), latest)
    # DÜZELTME: Satış hedefi ARTIK ÖNCELİKLE Excel'deki "Satış Hedefi" sayfasından
    # okunuyor (Ay, MagazaID, Hedef Ciro sütunları) — çünkü gerçek kullanım akışı
    # "Excel yükle -> Railway motoru hesaplar" şeklinde, Vercel'de ayrıca elle hedef
    # girilmesini beklemek bu akışla uyuşmuyordu. Vercel/Supabase girişi (varsa)
    # yalnız Excel'de o mağaza/ay için değer YOKSA yedek olarak kullanılır; açıklama/
    # aksiyon planı/sorumlu alanları hâlâ yalnız Vercel'den gelir (Excel'de karşılığı yok).
    excel_targets = _indexed(
        sheets.get("Satış Hedefi", sheets.get("Satis Hedefi", sheets.get("Satış Hedefi ", pd.DataFrame()))),
        latest,
    )
    target_frame = pd.DataFrame(targets)
    if not target_frame.empty:
        target_frame = target_frame[target_frame.get("period", "").astype(str).eq(latest)].copy()
        target_frame["_store_key"] = target_frame.get("store_id", "").astype(str).str.strip()
        target_frame = target_frame.drop_duplicates("_store_key", keep="last").set_index("_store_key")

    rows: list[dict[str, Any]] = []
    for _, store in stores.iterrows():
        store_id = str(store.get("MağazaID") or store.get("Mağaza") or "").strip()
        store_name = str(store.get("Mağaza") or store_id).strip()
        source_key = store_id if store_id in current.index else store_name
        target_key = store_id if not target_frame.empty and store_id in target_frame.index else store_name
        revenue = _value(current, source_key, "Aylık Ciro", "Ciro")
        tickets = _value(current, source_key, "Aylık Fiş", "Fiş Adedi")
        basket = _value(current, source_key, "Ort. Sepet", "Ortalama Sepet")
        revenue_change = _change(revenue, _value(prior, source_key, "Aylık Ciro", "Ciro"))
        excel_key = store_id if store_id in excel_targets.index else store_name
        target = _value(excel_targets, excel_key, "Hedef Ciro", "Satış Hedefi")
        if target is None:
            target = _value(target_frame, target_key, "sales_target")
        norm = _number(store.get("Norm", store.get("Norm Kadro"))) or 0
        current_staff = _number(store.get("Mevcut", store.get("Aktif Mevcut"))) or 0
        row = {
            "MağazaID": store_id, "Mağaza": store_name,
            "Norm": int(norm), "Mevcut": int(current_staff),
            "Norm Karşılama %": current_staff / norm * 100 if norm else 0,
            "Satış Hedefi": target, "Gerçekleşen Ciro": revenue,
            "Hedef Gerçekleşme %": revenue / target * 100 if revenue is not None and target not in (None, 0) else None,
            "Fiş Değişim %": _change(tickets, _value(prior, source_key, "Aylık Fiş", "Fiş Adedi")),
            "Sepet Değişim %": _change(basket, _value(prior, source_key, "Ort. Sepet", "Ortalama Sepet")),
            "Reel Büyüme %": ((1 + revenue_change / 100) / (1 + inflation_pct / 100) - 1) * 100 if revenue_change is not None else None,
            "Fazla Mesai Saat": _value(overtime, source_key, "Fazla Mesai Saat"),
            "Kayıp FTE": _value(absence, source_key, "Fiili Kayıp FTE"),
            "İş Yükü Endeksi": _value(workload, source_key, "İş Yükü Endeksi"),
            "Fire Oranı %": _value(waste, source_key, "Fire Oranı %"),
            "Yönetici Puanı": _value(performance, source_key, "Yönetici Puanı"),
            "Satış Açıklaması": target_frame.loc[target_key].get("explanation") if not target_frame.empty and target_key in target_frame.index else None,
            "Aksiyon Planı": target_frame.loc[target_key].get("action_plan") if not target_frame.empty and target_key in target_frame.index else None,
            "Sorumlu": target_frame.loc[target_key].get("owner_name") if not target_frame.empty and target_key in target_frame.index else None,
        }
        row["Otomatik Kök Neden"], row["Personel İddiası"], row["İstenen Aksiyon"] = _diagnose(row)
        rows.append(row)
    return pd.DataFrame(rows), latest, previous

File: vercel-ui/lib/sales-root-cause.ts
export type DataRow = Record<string, unknown>

export type SalesEvidence = {
  period: string
  storeId: string
  storeName: string
  revenue: number | null
  tickets: number | null
  basket: number | null
  revenueChange: number | null
  ticketChange: number | null
  basketChange: number | null
  realGrowth: number | null
  overtimeHours: number | null
  absenceDays: number | null
  lostFte: number | null
  workloadIndex: number | null
  wasteRate: number | null
  managerScore: number | null
  customerExperience: number | null
  onlineOrders: number | null
  goodsReceipt: number | null
}

export type RootCauseResult = {
  code: string
  title: string
  severity: 'good' | 'watch' | 'bad' | 'missing'
  evidence: string[]
  personnelClaim: 'supported' | 'unsupported' | 'inconclusive'
  requiredAction: string
}

const numberValue = (value: unknown): number | null => {
  if (value === null || value === undefined || value === '') return null
  if (typeof value === 'number') return Number.isFinite(value) ? value : null
  const parsed = Number(String(value).replace(/\s/g, '').replace(/\./g, '').replace(',', '.'))
  return Number.isFinite(parsed) ? parsed : null
}

export const textValue = (value: unknown) => String(value ?? '').trim()
export const rowPeriod = (row: DataRow) => textValue(row.Ay ?? row.Dönem ?? row.Donem ?? row.Tarih).slice(0, 7)
export const rowStoreId = (row: DataRow) => textValue(row.MagazaID ?? row['MağazaID'] ?? row.Mağaza ?? row.Magaza)
export const rowStoreName = (row: DataRow) => textValue(row.Mağaza ?? row.Magaza ?? row.MagazaID ?? row['MağazaID'])
export const pctChange = (current: number | null, previous: number | null) => current !== null && previous !== null && previous !== 0 ? ((current / previous) - 1) * 100 : null

const byStorePeriod = (rows: DataRow[], period: string) => new Map(
  rows.filter((row) => !period || !rowPeriod(row) || rowPeriod(row) === period).map((row) => [rowStoreId(row), row]),
)

export function buildSalesEvidence(args: {
  operations: DataRow[]
  overtime: DataRow[]
  absence: DataRow[]
  productivity: DataRow[]
  wasteReturns: DataRow[]
  performance: DataRow[]
  onlineOrders: DataRow[]
  goodsReceipt: DataRow[]
  inflationPct: number
}): { latestPeriod: string; previousPeriod: string; byStore: Map<string, SalesEvidence> } {
  const periods = [...new Set(args.operations.map(rowPeriod).filter(Boolean))].sort()
  const latestPeriod = periods.at(-1) ?? ''
  const previousPeriod = periods.at(-2) ?? ''
  const current = byStorePeriod(args.operations, latestPeriod)
  const previous = byStorePeriod(args.operations, previousPeriod)
  const overtime = byStorePeriod(args.overtime, latestPeriod)
  const absence = byStorePeriod(args.absence, latestPeriod)
  const productivity = byStorePeriod(args.productivity, '')
  const waste = byStorePeriod(args.wasteReturns, latestPeriod)
  const performance = byStorePeriod(args.performance, latestPeriod)
  const online = byStorePeriod(args.onlineOrders, latestPeriod)
  const goods = byStorePeriod(args.goodsReceipt, latestPeriod)
  const result = new Map<string, SalesEvidence>()
  current.forEach((row, storeId) => {
    const prior = previous.get(storeId)
    const revenue = numberValue(row['Aylık Ciro'] ?? row.Ciro)
    const tickets = numberValue(row['Aylık Fiş'] ?? row['Fiş Adedi'])
    const basket = numberValue(row['Ort. Sepet'] ?? row['Ortalama Sepet'])
    const previousRevenue = numberValue(prior?.['Aylık Ciro'] ?? prior?.Ciro)
    const revenueChange = pctChange(revenue, previousRevenue)
    result.set(storeId, {
      period: latestPeriod, storeId, storeName: rowStoreName(row), revenue, tickets, basket,
      revenueChange,
      ticketChange: pctChange(tickets, numberValue(prior?.['Aylık Fiş'] ?? prior?.['Fiş Adedi'])),
      basketChange: pctChange(basket, numberValue(prior?.['Ort. Sepet'] ?? prior?.['Ortalama Sepet'])),
      realGrowth: revenueChange === null ? null : (((1 + revenueChange / 100) / (1 + args.inflationPct / 100)) - 1) * 100,
      overtimeHours: numberValue(overtime.get(storeId)?.['Fazla Mesai Saat']),
      absenceDays: numberValue(absence.get(storeId)?.['Devamsızlık Gün']),
      lostFte: numberValue(absence.get(storeId)?.['Fiili Kayıp FTE']),
      workloadIndex: numberValue(productivity.get(storeId)?.['İş Yükü Endeksi']),
      wasteRate: numberValue(waste.get(storeId)?.['Fire Oranı %']),
      managerScore: numberValue(performance.get(storeId)?.['Yönetici Puanı']),
      customerExperience: numberValue(performance.get(storeId)?.['Müşteri Deneyimi']),
      onlineOrders: numberValue(online.get(storeId)?.['Günlük Sipariş']),
      goodsReceipt: numberValue(goods.get(storeId)?.['Günlük Mal Kabul']),
    })
  })
  return { latestPeriod, previousPeriod, byStore: result }
}

export function diagnoseSales(args: { evidence?: SalesEvidence; normRate: number | null; salesRate: number | null }): RootCauseResult {
  const { evidence: e, normRate, salesRate } = args
  if (!e || salesRate === null) return { code: 'VERI_EKSIK', title: 'Analiz için hedef veya gerçekleşen veri eksik', severity: 'missing', evidence: [], personnelClaim: 'inconclusive', requiredAction: 'Satış hedefi ve dönem verisini tamamla.' }
  if (salesRate >= 100 && (normRate ?? 100) < 95) return { code: 'EKSIK_KADROYLA_HEDEF_USTU', title: 'Eksik kadroyla hedef aşıldı', severity: 'watch', evidence: [`Hedef gerçekleşme ${salesRate.toFixed(1)}%`, `Norm karşılama ${(normRate ?? 0).toFixed(1)}%`], personnelClaim: 'unsupported', requiredAction: 'Sürdürülebilirlik, fazla mesai ve çalışan yükünü incele.' }
  if (salesRate >= 100) return { code: 'HEDEF_TUTTU', title: 'Satış hedefi gerçekleşti', severity: 'good', evidence: [`Hedef gerçekleşme ${salesRate.toFixed(1)}%`], personnelClaim: 'unsupported', requiredAction: 'Sonucu koru; fire ve iş yükü guardrail’lerini izle.' }

  const evidence: string[] = [`Hedef gerçekleşme ${salesRate.toFixed(1)}%`]
  if (e.ticketChange !== null) evidence.push(`Fiş değişimi ${e.ticketChange.toFixed(1)}%`)
  if (e.basketChange !== null) evidence.push(`Sepet değişimi ${e.basketChange.toFixed(1)}%`)
  if (e.realGrowth !== null) evidence.push(`Reel büyüme ${e.realGrowth.toFixed(1)}%`)
  if (e.wasteRate !== null) evidence.push(`Fire ${e.wasteRate.toFixed(1)}%`)
  if ((normRate ?? 100) >= 98) return { code: 'TAM_KADRO_DUSUK_SATIS', title: 'Norm dolu; satış sapması kadroyla açıklanamaz', severity: 'bad', evidence, personnelClaim: 'unsupported', requiredAction: 'Fiş, sepet, stok/fire, kategori ve mağaza yönetimi aksiyonu iste.' }
  if ((e.ticketChange ?? 0) < -3 && (e.basketChange ?? 0) >= -3) return { code: 'MUSTERI_TRAFIGI', title: 'Ana sinyal: fiş/müşteri trafiği düşüşü', severity: 'bad', evidence, personnelClaim: 'unsupported', requiredAction: 'Trafik, kampanya, rekabet ve mağaza çekiciliği kök nedenini açıkla.' }
  if ((e.basketChange ?? 0) < -3) return { code: 'SEPET_DUSUSU', title: 'Ana sinyal: ortalama sepet düşüşü', severity: 'bad', evidence, personnelClaim: 'unsupported', requiredAction: 'Ürün karması, fiyat, kampanya ve stok bulunurluğunu açıkla.' }
  const peopleEvidence = (normRate ?? 100) < 95 && ((e.overtimeHours ?? 0) > 0 || (e.lostFte ?? 0) > 0 || (e.workloadIndex ?? 0) >= 70)
  if (peopleEvidence) return { code: 'PERSONEL_ETKISI_KANITLI', title: 'Personel etkisi veriyle destekleniyor', severity: 'bad', evidence, personnelClaim: 'supported', requiredAction: 'Eksik unvan/saat ile satış kaybı bağlantısını ve telafi planını kaydet.' }
  return { code: 'KOK_NEDEN_ACIKLANMADI', title: 'Satış sapmasının kök nedeni açıklanmadı', severity: 'bad', evidence, personnelClaim: 'inconclusive', requiredAction: 'Fiş, sepet, stok, kategori ve yönetici kırılımıyla kanıt sun.' }
}
